### ISSUE TITLE
Strange behavior for dask.bag.fold

### ISSUE THREAD

**Issue by StuartAxelOwen** [ContinuumIO/dask#511]
Example:

`toolz.reduce`:
```python
>>> from toolz import reduce
>>> def num_cat(l, num):
...     return l + [num]
>>> reduce(num_cat, range(5), [])
[0, 1, 2, 3, 4]
```

But now, with `bag.fold`:
```python
>>> from dask import bag
>>> def num_cat(l, num):
...     return l + [num]
>>> bag.from_sequence(range(5)).fold(num_cat, initial=[]).compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stuart/code/python/dask/dask/bag/core.py", line 191, in compute
    return get(self.dask, self.key, **kwargs)
  File "/Users/stuart/code/python/dask/dask/bag/core.py", line 108, in get
    return get(dsk2, keys, **kwargs)
  File "/Users/stuart/code/python/dask/dask/multiprocessing.py", line 56, in get
    queue=queue, get_id=_process_get_id, **kwargs)
  File "/Users/stuart/code/python/dask/dask/async.py", line 474, in get_async
    _execute_task(task, data)  # Re-execute locally
  File "/Users/stuart/code/python/dask/dask/async.py", line 245, in _execute_task
    return func(*args2)
  File "<stdin>", line 2, in num_cat
TypeError: unsupported operand type(s) for +: 'int' and 'list'
```

Hmmm, maybe the params are switched for the reducer?

```python
>>> def num_cat(num, l):
...     return l + [num]
bag.from_sequence(range(5)).fold(num_cat2, initial=[]).compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stuart/code/python/dask/dask/bag/core.py", line 191, in compute
    return get(self.dask, self.key, **kwargs)
  File "/Users/stuart/code/python/dask/dask/bag/core.py", line 108, in get
    return get(dsk2, keys, **kwargs)
  File "/Users/stuart/code/python/dask/dask/multiprocessing.py", line 56, in get
    queue=queue, get_id=_process_get_id, **kwargs)
  File "/Users/stuart/code/python/dask/dask/async.py", line 474, in get_async
    _execute_task(task, data)  # Re-execute locally
  File "/Users/stuart/code/python/dask/dask/async.py", line 245, in _execute_task
    return func(*args2)
  File "<stdin>", line 2, in num_cat2
TypeError: unsupported operand type(s) for +: 'int' and 'list'
```

Nope... Seems like somewhere, somehow the params are being switched - perhaps when it goes from the combine to the actual reduction?  This passed tests because all `fold` tests used reducers in which parameter order doesn't matter.

**Comment by StuartAxelOwen** [comment]
Ah, I see - when executing the dask, `fold` seems to expect a reducer type signature of `(a, a) -> a`, as reducer results are fed back into the reducer when no combine operation is provided.

**Comment by StuartAxelOwen** [comment]
This looks like expected behavior - that would make it more of a documentation bug.

**Comment by mrocklin** [comment]
This was due to bad checking on the `initial` kwarg.  Fixed in #544 .  Closing.
